### ISSUE TITLE
[BUG] OpenSSL internal error: refcount error

### ISSUE THREAD

**Issue by XieJiSS** [trojan-gfw/trojan#130]
- [x] I certify that I acknowledge if I don't follow the format below, or I'm using an old version of trojan, or I apparently fail to provide sufficient information (such as logs, specific numbers), or I don't check this box, my issue will be closed immediately without any notice.

**Trojan Version** 1.12.3
The version of trojan you are using.
1.12.3

**Describe the bug**
After running for a period of time (~7 days), my server side trojan process exited with error log:
```
ssl/ssl_sess.c:785: OpenSSL internal error: refcount error
```

**To Reproduce**
1. Maybe keep server-side trojan process running for a long time?

**Expected behavior**
It should not quit unexpectedly

**Logs**
```
[2019-09-24 13:07:15] [WARN] 172.x.x.x:xxxx not trojan request, connecting to 127.0.0.1:80
[2019-09-24 13:08:21] [INFO] 172.x.x.x:xxxx disconnected, 7794 bytesreceived, 611 bytes sent, lasted for 66 seconds
ssl/ssl_sess.c:785: OpenSSL internal error: refcount error
```

**Environment**
Where are you running trojan? What is your proxy set up?
```
$ trojan -v
Welcome to trojan 1.12.3
Boost 1_70, OpenSSL 1.1.1c  28 May 2019
 [Enabled] MySQL Support (10.3.6)
 [Enabled] TCP_FASTOPEN Support
 [Enabled] TCP_FASTOPEN_CONNECT Support
 [Enabled] SSL KeyLog Support
```

Linux 4.15.0-1040 Ubuntu 18.04.3 LTS x86_64

**Additional context**
Nothing to provide here.

maybe related to: https://github.com/mpeylo/cmpossl/issues/54

**Comment by GreaterFire** [comment]
It's a rather rare and strange error. Will look into this. Might be an upstream bug.

**Comment by GreaterFire** [comment]
Seems that the reference count of an SSL session is decremented to a negative number.
https://github.com/openssl/openssl/blob/OpenSSL_1_1_1c/ssl/ssl_sess.c#L785

**Comment by wongsyrone** [comment]
What's the source of your trojan binary? Compiled by yourself or fetched directly from GitHub release page.

**Comment by wongsyrone** [comment]
BTW, please post the output of `ldd <path-to-your-trojan-binary>`

**Comment by GreaterFire** [comment]
Given the version output I think there is a high chance that the binary is my pre-built static binary.

**Comment by GreaterFire** [comment]
@JieJiSS Can you please provide the coredump (by coredumpctl if you use systemd) of that run if possible (because the error causes a SIGABRT and produce a coredump). Also, please provide the ldd result as stated by @WongSyrone above.

**Comment by GreaterFire** [comment]
@JieJiSS Is it possible for you to follow up on this issue?

**Comment by XieJiSS** [comment]
@wongsyrone 

```shell
$ ldd /usr/local/bin/trojan
        linux-vdso.so.1 (0x00007ffeed1ed000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f1fd2126000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f1fd1d9d000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f1fd19ff000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f1fd17e7000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1fd15c8000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1fd11d7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1fd232a000)
```

**Comment by XieJiSS** [comment]
I'm sorry, but after some careful investigation, I realized that the coredump of this crash wasn't saved on my disk because I neither enabled the Ubuntu system's `Apport` nor started `trojan` via `systemctl`.

I've migrated to `systemctl` now, and if the same crash occurred again, I'll report to you asap.

Since I cannot provide any further information, I think we can close this issue for now.

Apologize for the inconvenience again.
